### PR TITLE
Channel selection overlay

### PR DESCRIPTION
**Pull request by DrabWeb**
screen shot 2017-05-26 at 3 52 19 am

Only supports joining at the moment

### CODE REVIEW

File: osu.Game/Overlays/Chat/ChannelListItem.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                channel = value;
+
+                name.Text = Channel.ToString();
+                topic.Text = Channel.Topic;
+                channel.Joined.ValueChanged += updateColour;
+                updateColour(Channel.Joined);
+            }
+        }
+
+        public ChannelListItem()
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: make this accept a channel in the ctor IMO.

also make a local bindable<bool> and use bindto in this scenario.

File: osu.Game/Overlays/Chat/ChannelListItem.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            {
+                FadeTo(value ? 1f : 0f, 100);
+            }
+        }
+
+        public Action<Channel> OnRequestJoin;
+        public Action<Channel> OnRequestLeave;
+
+        private Channel channel;
+        public Channel Channel
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: just remove this property. i can't think of a case we'd want to change the channel being displayed by a `ChannelListItem`

File: osu.Game/Overlays/Chat/ChannelListItem.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                updateColour(Channel.Joined);
+            }
+        }
+
+        public ChannelListItem()
+        {
+            RelativeSizeAxes = Axes.X;
+            AutoSizeAxes = Axes.Y;
+
+            Action = () => { (Channel.Joined ? OnRequestLeave : OnRequestJoin)?.Invoke(Channel); };
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: I'd probably make a named `onClick` method for this just for clarity.

File: osu.Game/Overlays/ChatOverlay.cs [hunk 4]
```diff
@@ -234,6 +244,16 @@ private void initializeChannels()
                     addChannel(channels.Find(c => c.Name == @"#lazer"));
                     addChannel(channels.Find(c => c.Name == @"#osu"));
                     addChannel(channels.Find(c => c.Name == @"#lobby"));
+
+                    channelSelection.OnRequestJoin = addChannel;
+                    channelSelection.Sections = new[]
+                    {
+                        new ChannelSection
+                        {
+                            Header = @"ALL CHANNELS",
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: no @ prefix. also for cases of upper-case, it's best to use `.ToUpper` and leave the actual string in title case I think.

File: osu.Game/Overlays/Chat/ChannelListItem.cs (hunk 1 above)
Comment (DrabWeb) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: For `Joined`?

File: osu.Game/Overlays/Chat/ChannelListItem.cs (hunk 3 above)
Comment (DrabWeb) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: Not exactly sure what you mean here

**Comment by peppy** [comment]
As discussed in discord, we need to make this hidden by default. Let's also make it stretch to take up all available space for the time being. If the chat area is resized to the point the channel selector becomes useless (ie. it can't display 5 channels or so) then it should automatically close.

**Review by peppy — DISMISSED** [review]
as mentioned

**Comment by peppy** [comment]
https://streamable.com/lc39r

**Comment by peppy** [comment]
Also this feels a bit janky.. would be nicer if it transitioned to the new size https://streamable.com/movy4

File: osu.Game/Overlays/Chat/ChannelListItem.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        private const float text_size = 15;
+        private const float transition_duration = 100;
+
+        private readonly Channel channel;
+
+        private readonly OsuSpriteText name;
+        private readonly OsuSpriteText topic;
+        private readonly TextAwesome joinedCheckmark;
+
+        private Color4? joinedColour;
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: why nullable?

File: osu.Game/Overlays/Chat/ChannelListItem.cs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+        protected override void OnHoverLost(InputState state)
+        {
+            if (!channel.Joined.Value)
+                name.FadeColour(Color4.White, transition_duration);
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            if(channel != null) channel.Joined.ValueChanged -= updateColour;
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChannelListItem.cs]: can channel ever be null?

**Comment by peppy** [comment]
- Was it always possible to see the channel overlay appearing behind the semi-transparent tab bar? we need to fix that.
- The tab should probably change state visually when toggled on

**Comment by peppy** [comment]
Opening channel select then hitting escape doesn't restore the state of the button :(

windows_10

File: osu.Game/Overlays/Chat/ChatTabControl.cs [hunk 7]
```diff
@@ -207,11 +207,12 @@ public class ChannelSelectorTabItem : ChannelTabItem
             {
                 public override bool Active
                 {
-                    get { return base.Active; }
+                    get { return false; }
                     set
                     {
-                        activeBindable.Value = value;
-                        base.Active = value;
+                        activeBindable.Value = !activeBindable.Value;
```
Comment (peppy) [inline osu.Game/Overlays/Chat/ChatTabControl.cs]: this is soooo hacky...

**Comment by peppy** [comment]
I'm merging this in under the premise that I intend to fix the add-new tab this week, by adding a new `ToggleableTab` type and avoiding the hacky implementation.

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
Remaining issue: Clicking channel tabs while the selection overlay is open causes no updates. May be a regression from my changes, needs further investigation.

**Review by peppy — APPROVED** [review]

**Pull request merged**